Check the return value of critical OpenSSL functions

toru opened the pull request:
## Objective

~Assert that `RAND_bytes()` and `EVP_CIPHER_CTX_cleanup()` were called successfully.~

Abort if `RAND_bytes()` and `EVP_CIPHER_CTX_cleanup()` ever fails.

## Why

`RAND_bytes()` is a fundamental cryptographic building block of the library, hence we should not assume that it will always succeed. Ditto for cleaning up the cipher context.

## Define Success

Here's an except From the OpenSSL documentation.

> RAND_bytes() and RAND_priv_bytes() return 1 on success, -1 if not supported by the current RAND method, or 0 on other failure.

> EVP_CIPHER_CTX_cleanup() returns 1 for success and 0 for failure.

Thanks to David Wong for reporting the concern.

deweerdt commented on lib/openssl.c:
```diff
@@ -79,7 +79,8 @@ static void HMAC_CTX_free(HMAC_CTX *ctx)
 {
-    RAND_bytes(buf, (int)len);
+    int ret = RAND_bytes(buf, (int)len);
+    assert(ret);
```
I believe this should be `!= 1`? 

> RAND_bytes() and RAND_priv_bytes() return 1 on success, -1 if not supported by the current RAND method, or 0 on other failure.

toru replied:
Thank you! Agreed that I made a bad assumption. I will assert on `== 1` since that's the success case.

toru commented:
@kazuho Thank you for the prompt review. Agreed that it's better to fail loud and hard for these failures. I've addressed your feedback in: eeb3bdc1d79d347c74d37a98de5a1d24fed9fe7f

toru commented:
Thank you for the pointers. It slipped from my mind that the cleanup is indeed internal to the file. I've made an amend based on your feedback + moved up the `cleanup_cipher_ctx` function to align with other static functions -- 62bcdfc19bba0fba4c8fae39c543a01c68c17a11.

toru commented:
Thank you for the pointers. It slipped from my mind that the cleanup is indeed internal to the file. I've made an amend based on your feedback + moved up the `cleanup_cipher_ctx` function to align with other static functions -- 62bcdfc19bba0fba4c8fae39c543a01c68c17a11.

toru commented:
Thank you for the pointers. It slipped from my mind that the cleanup is indeed internal to the file. I've made an amend based on your feedback + moved up the `cleanup_cipher_ctx` function to align with other static functions -- 62bcdfc19bba0fba4c8fae39c543a01c68c17a11.

toru commented:
Thank you for the pointers. It slipped from my mind that the cleanup is indeed internal to the file. I've made an amend based on your feedback + moved up the cleanup_cipher_ctx function to align with other static functions -- 62bcdfc

kazuho commented:
Thank you for the changes. Merging.
